### PR TITLE
refactoring

### PR DESCRIPTION
**Pull request by xgrommx**

### CODE REVIEW

File: src/Higher/Data/Foldable.purs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import Data.Monoid.Dual (Dual(..))
+import Data.Monoid.Endo (Endo(..))
+import Data.Newtype (unwrap)
+import Higher.Data.Functor (class HFunctor)
+import Higher.NaturalTransformation (type (~>.))
+
+class HFunctor h <= HFoldable (h :: (Type -> Type) -> (Type -> Type)) where
+  hfoldr :: forall a b. (a ~>. (b -> b)) -> b -> (h a ~>. b)
+  hfoldl :: forall a b. (b -> (a ~>. b)) -> b -> (h a ~>. b)
+  hfoldMap :: forall f m a. Monoid m => (f ~>. m) -> h f a -> m
```
Comment (natefaubion) [inline src/Higher/Data/Foldable.purs]: If Monads are Monoids in the category of Endofunctors, should hfoldMap have a Monad constraint instead? 😝

Comment (natefaubion) [reply src/Higher/Data/Foldable.purs]: In all seriousness, is there some prior art for this. To me `hfoldMap :: forall f m a. Monad m => (f ~> n) -> h f ~> m` looks like a pretty reasonable signature.

Comment (natefaubion) [reply src/Higher/Data/Foldable.purs]: Or even `Alternative`, which is more like a Monoid for Functors.

Comment (natefaubion) [reply src/Higher/Data/Foldable.purs]: `hfoldMap` with `Monad` would also mean interpreting `HFree` is just `hfoldMap` 😱

Comment (xgrommx) [reply src/Higher/Data/Foldable.purs]: I'm not sure that `hfoldMap` should has constraint `Monad`

Comment (xgrommx) [reply src/Higher/Data/Foldable.purs]: You can take a look on `compdata`

**Pull request merged**